typing/typecore.ml

```diff
@@ -4543,10 +4756,12 @@ and type_application env app_loc (mode : expected_mode) funct sargs =
               raise(Error(funct.exp_loc, env, Apply_non_function
                             (expand_head env funct.exp_type)))
     in
+    let m_arg = Alloc_mode.newvar () in
+    Alloc_mode.submode_exn m_arg m_param;
```

Comment:
This only matters for known partial applications, which now don't complain about capturing the argument unless it is actually local.

Reply:
Ah, I hadn't considered the partial case, that makes sense and is an improvement in behaviour.

I was slightly worried about the large number of new variables being created by this change. I checked, and this change increases the number of mode variables in a build of the compiler by about 40% ([results here](https://gist.github.com/stedolan/eb55c41680276d6390ed41e35387c722)). This doesn't seem to be much of a problem, but if we later run into problems with the mode variable solver this might be something to revisit.